src/Native/Unix/System.Native/pal_random.cpp

```diff
@@ -20,72 +20,51 @@ extern "C" void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* b
 {
     assert(buffer != NULL);
 
-    int rand_des;
-    char buf;
+    static volatile int rand_des = -1;
     long num = 0;
-
-    static bool sMissingDevRandom;
     static bool sMissingDevURandom;
     static bool sInitializedMRand;
 
-    int i = 0;
-
-    if (i < bufferLength && !sMissingDevRandom)
+    if (!sMissingDevURandom)
     {
-        // request non-blocking access to avoid hangs if the /dev/random is exhausted
-        // or just simply broken
-        if ((rand_des = open("/dev/random", O_RDONLY | O_NONBLOCK)) == -1)
+        if (rand_des == -1)
         {
-            if (errno == ENOENT)
+            int fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+            if (fd != -1)
             {
-                sMissingDevRandom = true;
-            }
-
-            // Back off and try /dev/urandom.
-        }
-        else
-        {
-            for( ; i < bufferLength; i++)
-            {
-                if (read(rand_des, &buf, 1) < 1)
+                if (!__sync_bool_compare_and_swap(&rand_des, -1, fd))
```

Comment:
How many handles does have a small .NET process opened  (e.g. `Console.ReadLine`)? Just want to understand the relative cost of one more permanently open handle.

Reply:
There are 21 handles open without my change, so that would be 23 with it (one in corefx, one in coreclr)

Reply:
Ok, it does not sound that bad.